### PR TITLE
Update storage and mint new wrapped ERC20s for a single validator node

### PR DESCRIPTION
**Pull request by james-chf**
Closes https://github.com/anoma/namada/issues/144

Closes https://github.com/anoma/namada/issues/140

This PR makes it so that we apply a derived tx whenever we apply a `ProtocolTxType::EthereumEvents`. This derived tx:
- for each event, writes the event to storage if it is the first time is being seen
- for any `TransferToNamada` event which becomes `seen`, mints wrapped ERC20s

This PR is implementing the previous design of the Ethereum bridge where we would have had vote extensions.

For this PR, Ethereum updates are only written to storage and acted upon the first time they are seen. In the degenerate case of a chain with a single validator this works, as the validator has 100% of the voting power so Ethereum events are always acted upon as soon as they are seen. Making the state update transaction work across multiple rounds and with multiple validators (https://github.com/anoma/namada/issues/197) will come in a later PR.

### CODE REVIEW

File: apps/src/lib/node/ledger/shell/mod.rs
Comment (james-chf) [inline apps/src/lib/node/ledger/shell/mod.rs:670]: TODO: `wasm_dir` is no longer necessary. `storage` should still be able to be passed as immutable during a dry_run_tx or query.

Comment (james-chf) [reply apps/src/lib/node/ledger/shell/mod.rs]: Being fixed up in https://github.com/anoma/namada/pull/310

**Comment by james-chf** [comment]
pls update wasm

**Comment by sug0** [comment]
Also closes https://github.com/anoma/namada/issues/127

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs
Comment (sug0) [inline apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs:56]: can you capitalize the first word of each of these docstrings?

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs
Comment (sug0) [inline apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs:32]: you can pick a `HashSet` or `BTreeMap` here, even if we need deterministic data in storage. borsh should guarantee we get the same encoding every time.

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs
Comment (sug0) [inline apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs:87]: you should be getting these based on some block height param, especially with the new shimming stuff

Comment (sug0) [reply apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs:87]: to be honest, I'm not entirely sure on this

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs
Comment (sug0) [inline apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs:190]: is it possible to split each of these branches of the if expr into two different helper funs?

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs
Comment (sug0) [inline apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs:261]: use `FractionalVotingPower::TWO_THIRDS`

File: apps/src/lib/node/ledger/protocol/transactions/store.rs
Comment (sug0) [inline apps/src/lib/node/ledger/protocol/transactions/store.rs:30]: tomas is adding a `StoreRead` and `StoreWrite` set of traits, soon, I think

**Review by sug0 — CHANGES_REQUESTED** [review]

File: apps/src/lib/node/ledger/protocol/transactions/store.rs
Comment (batconjurer) [inline apps/src/lib/node/ledger/protocol/transactions/store.rs:119]: Why use this when we have the `Storage<MockDB, Sha256Hasher>` for testing?

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/events.rs
Comment (batconjurer) [inline apps/src/lib/node/ledger/protocol/transactions/ethereum_events/events.rs:13]: doc string please

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs
Comment (batconjurer) [inline apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs:71]: doc string please

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/utils.rs
Comment (batconjurer) [inline apps/src/lib/node/ledger/protocol/transactions/ethereum_events/utils.rs:79]: use the assert_matches! macro

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/utils.rs
Comment (batconjurer) [inline apps/src/lib/node/ledger/protocol/transactions/ethereum_events/utils.rs:138]: assert_matches!

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/events.rs
Comment (james-chf) [reply apps/src/lib/node/ledger/protocol/transactions/ethereum_events/events.rs:20]: We shouldn't need to look at the nonce when processing `TransferToNamada` events, as they can be processed in any order

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/events.rs
Comment (james-chf) [reply apps/src/lib/node/ledger/protocol/transactions/ethereum_events/events.rs:70]: Yeah, I'm not sure whether it is better to use `let _` or `_` here

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs
Comment (james-chf) [reply apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs:32]: I agree, we should use a `BTreeSet`. I would rather make this change in a new PR though (and update the specs at the same time), I can make an issue for it

Comment (james-chf) [reply apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs:32]: https://github.com/anoma/namada/issues/434

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/events.rs
Comment (james-chf) [reply apps/src/lib/node/ledger/protocol/transactions/ethereum_events/events.rs:70]: I wanted to make explicit that we can ignore the result of `.insert` i.e. whether the supply key is being written for the first time, or whether it is being updated from a previous value, is not important at this point in the code

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs
Comment (james-chf) [reply apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs:87]: oh yeah good catch, this needs to change now. Leaving this comment thread open and will come back to it as need to think about it

File: apps/src/lib/node/ledger/protocol/transactions/store.rs
Comment (james-chf) [reply apps/src/lib/node/ledger/protocol/transactions/store.rs:30]: These two issues?
https://github.com/anoma/namada/issues/166

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs
Comment (james-chf) [reply apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs:188]: We work out the fractional voting power in the function itself (using `total_voting_power` as the denominator), though not using the `FractionalVotingPower` type (yet). I'll try to make this clearer the voting powers here are absolute and should be for the same epoch i.e.

`total_voting_power` - total voting power of all the active validators for epoch `n`
`voting_powers` - the absolute voting power for all active validators for epoch `n` who submitted an Ethereum event in their vote extension

**Comment by james-chf** [comment]
I will try updating to use our `FractionalVotingPower` type whereever possible (i.e. where this PR was using `Ratio<u64>` or `(u64, u64)` before

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs
Comment (batconjurer) [reply apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs:188]: That part is fine, but when I look at the unimplemented calculate_diff function, we don't have all the necessary info being passed in. This is because the previous voting powers may have come from a different epoch, but I only see the total_voting_power for the current epoch being passed int.

Comment (batconjurer) [reply apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs:188]: And to get that data, it think it is easier to change the signatures all the way up and use normalized voting powers (i.e. fractional voting powers)

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs
Comment (james-chf) [reply apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs:94]: So in the previous design with vote extensions, using whoever the active validators were for the most recent committed block made sense, as the Ethereum events would have been voted for during the consensus round for that block. e.g. in the consensus round for proposing block `n`, validators for block `n` vote for Ethereum events. Then the Ethereum events protocol tx is made and applied in block `n+1`. The validator sets may be different between blocks `n` and `n+1` if we crossed into a new epoch, so we get the validators for block `n`. (I will try to make the code here clearer as well as all this is happening a bit implicitly)

I'm not so familiar yet with this part of the new design but that can be further issues/PRs to migrate this old code to using a specified block height (as provided in `ProtocolTxType::EthereumEvents`, I guess?). 

NB: I've now updated the PR description to make explicit this is for the old ABCI++ design.

**Comment by james-chf** [comment]
pls update wasm

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs
Comment (sug0) [reply apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs:68]: okay sounds good. maybe add a `TODO` here, though? so that we know we need to change this later when `abciplus` shims merge

File: apps/src/lib/node/ledger/protocol/transactions/store.rs
Comment (sug0) [reply apps/src/lib/node/ledger/protocol/transactions/store.rs:30]: likewise, add a `TODO`, so we remember to change this later

**Review by sug0 — CHANGES_REQUESTED** [review]

**Comment by sug0** [comment]
I left comments in two open threads, if it's not clear what this review is about

**Comment by james-chf** [comment]
Rebasing on top of the shim vext changes from https://github.com/anoma/namada/pull/437/files , will also add an integration test using the `Storage` component

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/events.rs
Comment (batconjurer) [reply apps/src/lib/node/ledger/protocol/transactions/ethereum_events/events.rs:73]: I didn't know this either because apparently my IDE marks it as a syntax error. :exploding_head:

**Review by batconjurer — CHANGES_REQUESTED** [review]
So I've spent some time thinking about it and I really would prefer not to have the `StoreTrait` here. If it were needed for mocking, I would except it as a necessary evil, but the `FakeStorage` is almost identical to using the existing `MockDB`.

So if we have a mock we don't really need, that leaves us with a trait implemented by a single type, which I'm not in favor of as a general practice. I think it bloats the code and adds cognitive load by adding a layer of indirection over the code flow.  This would be fine if it was an api facing outward from a third party library, but since this is internal, I don't see it as a realistic goal to hide / abstract details away that we should actually be aware of.

This is a divisive issue in software development ( as it touches on the role of encapsulation and how much OOP should be allowed in a codebase ), so in order to resolve this philosophical dispute equitably, I have requested that Heliax define a set of code standards that everyone will be able to give their input on. This should happen some in the next weeks. 

As such, we should put merging this on hold until we have a clear set of guidelines on how to proceed. I will request changes for now to reflect this, and if need be I will change it to approve if it is decided this particular coding practice is acceptable.

**Comment by james-chf** [comment]
> So I've spent some time thinking about it and I really would prefer not to have the `StoreTrait` here. If it were needed for mocking, I would except it as a necessary evil, but the `FakeStorage` is almost identical to using the existing `MockDB`.

> So if we have a mock we don't really need, that leaves us with a trait implemented by a single type, which I'm not in favor of as a general practice. I think it bloats the code and adds cognitive load by adding a layer of indirection over the code flow. This would be fine if it was an api facing outward from a third party library, but since this is internal, I don't see it as a realistic goal to hide / abstract details away that we should actually be aware of.

I think the layer of indirection and hiding away the details from the `ethereum_events` module is worth it (though it's more lines of code). There are some issues with injecting `Storage` directly, even though that would work as well. The main two things are:

- we don't need the majority of its API, only a few methods - read, write, has_key - so we are making this module depend on a large interface it doesn't need
- we couple our module to only being able to use that `Storage` struct, and not anything else

The read/write/has_key methods of `Storage` are not perfectly suited for the derived transaction logic, as they return a gas cost which isn't relevant in this module - so we would still want an adapter around the `Storage` imo to discard this (and provide the API we actually want). This is what the `LedgerStore` implementation does in this PR - discards irrelevant values like gas cost so we don't need to do so throughout the `ethereum_events` module.

Using the `Store` trait, we can provide any implementation of that trait which is really useful for tests (some to be written). So far this PR only has a simple `FakeStorage` but we could implement some other more involved stubs or fakes, and in theory we could autogenerate mocks even (like with [mockall](https://docs.rs/mockall/latest/mockall)). By injecting the `Storage` struct directly, we would have to make changes to the `Storage` struct if we wanted to do that (some of which might not be suitable for tests).

The fake `Storage<MockDB, Sha256Hasher>` would work in effect identically to the simple `FakeStorage` for the purposes of tests here (I think) but is a more complex component. That isn't a big deal though. I see it could be more practical to just use it rather than implementing `FakeStorage`. We should definitely have integration tests with a real `Storage` component anyway. 

The `Store` trait here as declared just for use by a single type seems a bit overkill, but it serves to explicitly document the  API that derived transactions want i.e. a simple read/write/has_key one without gas costs. Passing a `Storage` in directly, you need to look at the code in `ethereum_events` to see what parts of its API are actually being used rather than just looking at the trait.
 
> This is a divisive issue in software development ( as it touches on the role of encapsulation and how much OOP should be allowed in a codebase ), so in order to resolve this philosophical dispute equitably, I have requested that Heliax define a set of code standards that everyone will be able to give their input on. This should happen some in the next weeks.
> 
> As such, we should put merging this on hold until we have a clear set of guidelines on how to proceed. I will request changes for now to reflect this, and if need be I will change it to approve if it is decided this particular coding practice is acceptable.

It's not really to do with OOP, I think encapsulation is good though and it'd be good to have coding guidelines. The most important thing here (imo) is to inject only the interface that we need, for code clarity, and one that is suited to us. This should make the `ethereum_events` code simpler.

An equivalent struct with that API that the `Store` trait has (read/write/has_key) would at least do the job of interface segregation. Ideally the `Storage` struct could have some subcomponent which implemented a `Store`-like trait that we could pass in directly (and the `Storage` struct itself would use that internally also) - that would provide most of the benefits although we'd still be coupled to that one implementation.

I'll try making the changes requested - switch to using the `Storage` struct directly and the fake `Storage<MockDB, Sha256Hasher>` for testing, to compare, then we could merge that asap and come back to changing this if we want to later.

**Comment by james-chf** [comment]
I've updated to use `Storage`, going to readapt now based on the logic that came in for ABCI 0.18, then will retest. Specifically we can no longer just get active validators/voting powers for the epoch of the last committed block, but must look at individual `MultiSignedEthEvent` to get the appropriate validators/voting powers.

**Comment by james-chf** [comment]
pls update wasm

**Comment by james-chf** [comment]
pls update wasm

**Comment by james-chf** [comment]
This PR became a bit overgrown, there are a few small things that should be in separate PRs that aren't directly to do with what this PR is trying to achieve:
- the `impl Add`/`impl AddAssign` for `FractionalVotingPower`
- the extra debug logging when a validator votes for Ethereum events
- ensuring the `QueriesExt` trait is visible outside of the `node::ledger::shell` module (as we want to use it in the `node::ledger::protocol` module as well)

In terms of what this PR is concerned with, it's still too much, it could at least be split up into PRs for:
- adding the `EthMsg`/`EthMsgUpdate` types (`ethereum_events/eth_msgs.rs`)
- the logic for storing newly seen Ethereum events (most of `ethereum_events/mod.rs`)
- the logic for minting wrapped ERC20s (`ethereum_events/events.rs`)

**Comment by james-chf** [comment]
pls update wasm

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/update.rs
Comment (sug0) [inline apps/src/lib/node/ledger/protocol/transactions/ethereum_events/update.rs:12]:
```suggestion
    update: impl FnOnce(&mut Amount),
```
I think you can make this a `FnOnce`

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/update.rs
Comment (sug0) [inline apps/src/lib/node/ledger/protocol/transactions/ethereum_events/update.rs:29]:
```suggestion
    update: impl FnOnce(&mut T),
```
same here

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/utils.rs
Comment (sug0) [inline apps/src/lib/node/ledger/protocol/transactions/ethereum_events/utils.rs:49]: same thing here

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/utils.rs
Comment (sug0) [inline apps/src/lib/node/ledger/protocol/transactions/ethereum_events/utils.rs:60]: same thing here

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/eth_msgs.rs
Comment (james-chf) [reply apps/src/lib/node/ledger/protocol/transactions/ethereum_events/eth_msgs.rs:58]: This struct represents an event as stored under `/eth_msgs`, which has an explicit `seen` key which maps to this field. So as long as the stored event has a `seen` key, we should have this field here. Though it does seem kind of redundant in a way as it can be derived whether an event is seen or not based on the `voting_power` being >2/3 like you say. Maybe we should simplify our design and not store an explicit `seen` field at all?

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/read.rs
Comment (james-chf) [reply apps/src/lib/node/ledger/protocol/transactions/ethereum_events/read.rs:54]: Almost definitely, there is for certain functionality like this for VPs. This sort of wrapper logic is defined in a few places and we should consolidate

**Review by sug0 — APPROVED** [review]

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/read.rs
Comment (batconjurer) [inline apps/src/lib/node/ledger/protocol/transactions/ethereum_events/read.rs:17]:
```suggestion
    Ok(maybe_value(storage, key)?.unwrap_or_default())
```

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/utils.rs
Comment (batconjurer) [inline apps/src/lib/node/ledger/protocol/transactions/ethereum_events/utils.rs:130]:
```suggestion
        assert_matches!(
            voting_powers.get(&(sole_validator, BlockHeight(100))),
            Some(v) if *v == FractionalVotingPower::new(1, 1).unwrap()
        );
```

File: apps/src/lib/node/ledger/protocol/transactions/ethereum_events/utils.rs
Comment (batconjurer) [inline apps/src/lib/node/ledger/protocol/transactions/ethereum_events/utils.rs:211]:
```suggestion
        assert_matches!(
            voting_powers.get(&(validator_1, BlockHeight(100))),
            Some(v) if *v == FractionalVotingPower::new(100, 300).unwrap()
        );
        assert_matches!(
            voting_powers.get(&(validator_2, BlockHeight(100))), 
            Some(v) if *v == FractionalVotingPower::new(200, 300).unwrap()
        );
```

**Review by batconjurer — APPROVED** [review]
LGTM! There are a couple of small suggestions, but you can take them or leave them.

**Pull request merged**
